File: javascript/path_traversal/staticFiles.js
const express = require('express');
const fs = require('fs');
const path = require('path');

const router = express.Router();
const STATIC_DIR = '/var/www/static';

// VULN 1: fs.readdirSync with user-controlled directory - exposes arbitrary directory contents
router.get('/browse', (req, res) => {
    const subdir = req.query.dir || '';
    const target = path.join(STATIC_DIR, subdir);
    const entries = fs.readdirSync(target);
    res.json(entries);
});

// VULN 2: fs.readFileSync with template literal path - user controls name and type
router.get('/asset', (req, res) => {
    const name = req.query.name;
    const type = req.query.type;
    const filePath = `${STATIC_DIR}/${type}/${name}`;
    const content = fs.readFileSync(filePath);
    res.send(content);
});

// VULN 3: fs.copyFileSync with user-controlled destination - write to arbitrary path
router.post('/copy', (req, res) => {
    const { template, dest } = req.body;
    const src = path.join(STATIC_DIR, 'templates', template);
    fs.copyFileSync(src, dest);
    res.send('copied');
});

module.exports = router;

File: ruby/path_traversal/static_files.rb
require 'sinatra'
require 'json'
require 'fileutils'

STATIC_DIR = '/var/www/static'

# VULN 1: Dir.entries with user-controlled directory - exposes arbitrary directory contents
get '/browse' do
  subdir = params[:dir] || ''
  target = File.join(STATIC_DIR, subdir)
  Dir.entries(target).to_json
end

# VULN 2: File.read with string interpolation - user controls name and type
get '/asset' do
  name = params[:name]
  type = params[:type]
  File.read("#{STATIC_DIR}/#{type}/#{name}")
end

# VULN 3: FileUtils.cp with user-controlled destination - write to arbitrary path
post '/copy' do
  template = params[:template]
  dest = params[:dest]
  src = File.join(STATIC_DIR, 'templates', template)
  FileUtils.cp(src, dest)
  'copied'
end
